discord/ext/commands/core.py

```diff
@@ -1911,10 +1911,10 @@ async def predicate(ctx: Context[BotT]) -> bool:
         # if we're here, all checks failed
         raise CheckAnyFailure(unwrapped, errors)
 
-    return check(predicate)
+    return check(predicate)  # type: ignore
 
 
-def has_role(item: Union[int, str], /) -> Callable[[T], T]:
+def has_role(item: Union[int, str], /) -> Check:
```

Comment:
For `--verifytypes` these need to have their generics passed, something like `Check[Any]` works fine to silence it.

Reply:
I've gone ahead and added the generic Any where applicable.

```py
from typing_extensions import reveal_type
from discord.ext import commands


@commands.check(lambda ctx: True)
async def foo(ctx: commands.Context, x: int, y: str):
    ...

reveal_type(foo)  # Type of "foo" is "(ctx: Context[Unknown], x: int, y: str) -> Coroutine[Any, Any, None]"


@commands.check(lambda ctx: True)
@commands.command()
async def foo_com(ctx: commands.Context, x: int, y: str):
    ...


reveal_type(foo_com)  # Type of "foo_com" is "Command[None, (x: int, y: str), None]"
```

This is the typing of what you were interested in there, as well as the corresponding command (in case you wanted more information or inadvertently left a line out of what you wanted to see)

I also did a quick re-check after updating pyright, and noticed that

1. One of my local fixes (Specifically, an issue with the return type of `__call__` in the protocol) was not in the commit pushed. (what I tested had not reflected the commit 1:1)
   
3. The Generic type of a protocol needs to be covariant. (There can be arguments over that, but that's what the spec currently is, and this still works for subclassed context here)

The requested change as well as those two fixes have been rebased in, along with updating the base. I've made the commit message contain more details about the specific changes as well.